[FTQC] Add xz encoding method for Paulis

multiphaseCFD opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

[sc-91293]

This PR adds encode_xz related functions to a new `pauli_tacker` module.

Background

Paulis can be decomposed into $X$ and $Z$ together a phase information and it is referred as `xz_encoding` in this [ref](https://arxiv.org/abs/2103.02202).

A Pauli $P$ can be encoded as $encode_{xz}(P)=(x,z)=encode_{xz}(X^xZ^z)$. This means that:
 |$I$|$X$|$XZ$|$Z$|
|:-:|:-:|:-:|:-:|
 |$(0,0)$|$(1,0)$|$(1,1)$|$(0,1)$|

With this encoding, two Paulis can be multiplied up to global phase by xoring their encoded representation: $encode_{xz}(P_1P_2) = encode_{xz}(P_1) \oplus encode_{xz}(P_2) = (x_1 \oplus x_2, z_1 \oplus z_2)$


**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

joeycarter reviewed (commented):
Hey @multiphaseCFD, quick code-organization question before I continue reviewing. Do we want to put the new functions in ftqc/utils.py? Or would it be better to start a new module, maybe something like `pauli_encoding.py`, or more generally `pauli_frame_tracking.py` depending on what else we plan to include.

mlxd commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,74 @@
+    res = pauli_encode_xz(ops.pop())
+
+    while len(ops) > 0:
+        res = np.bitwise_xor(pauli_encode_xz(ops.pop()), res)
```
Should we offload to `qml.math` for this, instead of just numpy?

lillian542 reviewed (commented):
Thanks, @multiphaseCFD, looking great! 

I found it surprising initially that the function accepts operator _types_ rather than initialized operators; I initially expected it to accept things like`qml.X(0)` when reading. 

I think this makes a lot of sense as-is, since the full operator includes wire information that we would lose when we encode, and that's more clearly communicated by not _accepting_ the wire information in the first place. But it's a little unexpected compared to a lot of behaviour in PennyLane (outside of resource estimation). 

Can we update the docstring to give a very brief example,  so it's immediately clear when skimming it what input in accepts?

joeycarter commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,74 @@
+
+from typing import List, Tuple
+
+import pennylane as qml
```
```suggestion
from pennylane import I, X, Y, Z
from pennylane.operation import Operator
```

I believe the preference now is to avoid `import pennylane as qml` from within `pennylane`, and to prefer importing only the required classes/functions directly.

multiphaseCFD replied:
Done!

joeycarter commented on tests/ftqc/test_pauli_tracker.py:
```diff
@@ -0,0 +1,79 @@
+class TestPauliTracker:
+    """Test for the pauli tracker related functions."""
+
+    @pytest.mark.parametrize("op", [qml.I, qml.X, qml.Y, qml.Z, qml.S])
+    def test_pauli_encode_xz(self, op):
```
[Optional] Since the expected results are straightforward here, could we parameterize them next to the corresponding gate in the list of test inputs, e.g. something like

```suggestion
    @pytest.mark.parametrize("op, expected", [(qml.I, (0, 0)), (qml.X, (1, 0)), (qml.Y, (1, 1)), (qml.Z, (0, 1)), (qml.S, None)])
    def test_pauli_encode_xz(self, op, expected):
```

multiphaseCFD replied:
Done!

joeycarter reviewed (commented):
Thanks @multiphaseCFD, looks good! Just a few observations to add to the others.

multiphaseCFD commented on pennylane/ftqc/__init__.py:
```diff
@@ -61,4 +62,6 @@
     "convert_to_mbqc_gateset",
     "generate_lattice",
     "make_graph_state",
+    "pauli_encode_xz",
+    "pauli_prod_to_xz",
```
Sure

multiphaseCFD commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,74 @@
+_paulis = frozenset({qml.X, qml.Y, qml.Z, qml.I})
+
+
+def pauli_encode_xz(op: qml.operation.Operator) -> Tuple[np.uint8, np.uint8]:
```
Sure

multiphaseCFD commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,74 @@
+    while len(ops) > 0:
+        res = np.bitwise_xor(pauli_encode_xz(ops.pop()), res)
+
+    return res
```
Just update the return type from a tuple of int to a qml.operation.Operator or Paulis to make it safer.

multiphaseCFD commented on tests/ftqc/test_pauli_tracker.py:
```diff
@@ -0,0 +1,79 @@
+    def test_pauli_encode_xz(self, op):
+        if op not in [qml.I, qml.X, qml.Y, qml.Z]:
+            with pytest.raises(NotImplementedError):
+                _ = pauli_encode_xz(op)
```
Done!

multiphaseCFD commented on tests/ftqc/test_pauli_tracker.py:
```diff
@@ -0,0 +1,79 @@
+            ([qml.X, qml.Y, qml.Z, qml.I, qml.Z], (0, 1)),
+        ],
+    )
+    def test_pauli_prod_to_xz(self, ops, res):
```
Done

multiphaseCFD commented:
> Member

> I think this makes a lot of sense as-is, since the full operator includes wire information that we would lose when we encode, and that's more clearly communicated by not accepting the wire information in the first place. But it's a little unexpected compared to a lot of behaviour in PennyLane (outside of resource estimation).

yep, that is the reason why the functions accept operator types. Thanks for pointing out how to ensure the consistency between ftqc and the rest of ecosystem. I updated the `pauli_prod` and  `pauli_to_xz` method to accept the instances of Operators , while keep `xz_to_pauli` independent of wire information. 

>  Can we update the docstring to give a very brief example, so it's immediately clear when skimming it what input in accepts?

Updated

multiphaseCFD commented on doc/releases/changelog-dev.md:
```diff
@@ -571,6 +574,7 @@ Anton Naim Ibrahim,
 Lee J. O'Riordan,
 Mudit Pandey,
 Andrija Paurevic,
+Shuli, Shu
```
Nice catch! Thanks

multiphaseCFD commented:
> I noticed that when you removed the `pauli_tracker` imports from `ftqc/__init__.py`, the docs for the new functions aren't created (or at least they don't show up in the table of contents for the `ftqc` module docs). Sphinx must have a way to generate docs for entire submodules, although I'm not sure how to do that.

Thanks @joeycarter . Nice catch! Let me try to fix it in the commuting PR

joeycarter commented:
> > I noticed that when you removed the `pauli_tracker` imports from `ftqc/__init__.py`, the docs for the new functions aren't created (or at least they don't show up in the table of contents for the `ftqc` module docs). Sphinx must have a way to generate docs for entire submodules, although I'm not sure how to do that.
> 
> Thanks @joeycarter . Nice catch! Let me try to fix it in the commuting PR

I figured it out if you want to get it in here :slightly_smiling_face:. You can add the following lines to `doc/code/qml_ftqc.rst` (filing in the placeholders appropriately):

```rst
Pauli Tracker
^^^^^^^^^^^^^

.. currentmodule:: pennylane.ftqc.pauli_tracker

[Insert description of Pauli Tracker].

.. autosummary::
    :toctree: api

    pauli_to_xz
    xz_to_pauli
    pauli_prod
```

Tested locally and it looks good :heavy_check_mark:

joeycarter commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,137 @@
+        op (qml.operation.Operator): A Pauli operator.
+
+    Return:
+        A tuple of xz encoding data, :math:`x` is the exponent of the :class:`~pennylane.X`, :math:`z` is the exponent of
+    the :class:`~pennylane.Z`.
```
```suggestion
        A tuple of xz encoding data, :math:`x` is the exponent of the :class:`~pennylane.X`, :math:`z` is the exponent of
        the :class:`~pennylane.Z`.
```

Otherwise it gets put on a new line

joeycarter commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,137 @@
+    **Example:**
+        The following example shows how the XZ to Pauli works.
+
+        .. code-block:: python3
+            from pennylane.ftqc.pauli_tracker import xz_to_pauli
+            >>> xz_to_pauli(0, 0)
+            <class 'pennylane.ops.identity.Identity'>
```
Same here:

```suggestion
        .. code-block:: python3

            from pennylane.ftqc.pauli_tracker import xz_to_pauli
            >>> xz_to_pauli(0, 0)
            <class 'pennylane.ops.identity.Identity'>
```

joeycarter commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,137 @@
+
+def pauli_prod(ops: List[Operator]) -> Operator:
+    """
+    Get the result of a product of list of Pauli operators. The result is a new Pauli operator up to global phase.
```
```suggestion
    Get the result of a product of a list of Pauli operators. The result is a new Pauli operator up to a global phase.
```

joeycarter commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,137 @@
+    **Example:**
+        The following example shows how the `pauli_prod` works.
+
+        .. code-block:: python3
+            from pennylane.ftqc.pauli_tracker import pauli_prod
+            from pennylane import I, X, Y, Z
+            >>> pauli_prod([I(0),X(0),Y(0),Z(0)])
+            I(0)
```
Same here:

```suggestion
        .. code-block:: python3

            from pennylane.ftqc.pauli_tracker import pauli_prod
            from pennylane import I, X, Y, Z
            >>> pauli_prod([I(0),X(0),Y(0),Z(0)])
            I(0)
```

joeycarter commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,137 @@
+            >>> pauli_prod([I(0),X(0),Y(0),Z(0)])
+            I(0)
+
+        A Pauli operator is returned for a list of Pauli operator up to global phase.
```
```suggestion
        A Pauli operator is returned for a list of Pauli operator up to a global phase.
```

joeycarter reviewed (dismissed):
Just a few final comments and corrections on the docs, then we should be good to merge!

joeycarter commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,140 @@
+    **Example:**
+        The following example shows how the XZ to Pauli works.
+
+        .. code-block:: python3
+
+            from pennylane.ftqc.pauli_tracker import xz_to_pauli
+            >>> xz_to_pauli(0, 0)
+            <class 'pennylane.ops.identity.Identity'>
```
Possibly the docs build is failing because the output is not valid python code? Maybe using `pycon` instead will fix the issue? If we do that here we can update the other `code-block`s as well.

```suggestion
        .. code-block:: pycon

            >>> from pennylane.ftqc.pauli_tracker import xz_to_pauli
            >>> xz_to_pauli(0, 0)
            <class 'pennylane.ops.identity.Identity'>
```

multiphaseCFD replied:
Thanks @joeycarter! Let's return `xz_to_pauli(0, 0)(wires=0)` to get it around

lillian542 commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,140 @@
+        z (np.uint8) : Exponent of :class:`~pennylane.Z` in the Pauli record.
+
+    Return:
+        A Pauli operator.
```
```suggestion
        A Pauli operator class.
```

lillian542 commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,140 @@
+
+def xz_to_pauli(x: np.uint8, z: np.uint8) -> Operator:
+    """
+    Convert x, z to a Pauli operator.
```
```suggestion
    Convert x, z to a Pauli operator class.
```

lillian542 commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,140 @@
+            >>> xz_to_pauli(0, 0)
+            <class 'pennylane.ops.identity.Identity'>
+
+        A Pauli operator is returned for a given xz tuple.
```
```suggestion
        A Pauli operator class is returned for a given xz tuple.
```

lillian542 commented on pennylane/ftqc/pauli_tracker.py:
```diff
@@ -0,0 +1,140 @@
+
+        .. code-block:: python3
+
+            from pennylane.ftqc.pauli_tracker import pauli_prod
+            from pennylane import I, X, Y, Z
+            >>> pauli_prod([I(0),X(0),Y(0),Z(0)])
+            I(0)
```
Just to mention, for a list of Pauli operations, we do already have the option to do:

```
>>> qml.prod(I(0), X(0), Y(0), Z(0)).simplify()
1j * I(0)
```
I'm just wondering if that already suits our purposes. This implementation is faster though 😅

lillian542 commented on doc/releases/changelog-dev.md:
```diff
@@ -310,6 +310,9 @@
 * Alias for Identity (`I`) is now accessible from `qml.ops`.
   [(#7200)](https://github.com/PennyLaneAI/pennylane/pull/7200)
 
+* Add xz encoding related `pauli_encode_xz` and `pauli_prod_to_xz` functions to the `ftqc` module.
```
Function names are out-of-date in the change-log

lillian542 reviewed (approved):
Looks good to me! Just left a few very minor comments, but nothing else from me.
